File: R_package/R/rtgin1.R
rtgin1 <- function(a, m, sign, algo='hormann', verbose=FALSE){
  # If sign = True, draw from the positive region Z > 0
  # If sign = False, draw from the negative region Z < 0
  # Compute necessary values for ratio-of-uniforms method
  k <- sqrt(m^2 + 4 * a)
  if (sign) {
    mode <- (-m + k) / (2 * a)
    mult <- -1
  } else {
    mode <- (-m - k) / (2 * a)
    mult <- 1
  }

  if (algo == 'hormann') {
    algo_1 <- TRUE
  } else if (algo == 'leydold') {
    algo_1 <- FALSE
  } else {
    stop("algo must be either 'hormann' or 'leydold'")
  }

  # Draw from minimal bounding rectangle and accept draw
  vmax <- mshift(mode, mode, a, m, FALSE)
  if (algo_1) {
    # Hörmann and Leydold (2014) using Cardano method
    R <- polyroot(c(-mode, 1 + m * mode, -m + a * mode, 2 - a))
    roots <- Re(R)   #real part
    roots <- roots[(-mult*roots) > 0]
    uvals <- sapply(roots, mshift, mode = mode, a = a, m = m, shift = TRUE)
    uvals <- sort(uvals)
  } else {
    # Leydold (2001) using the proportionality constant
    lcons <- -0.25 * m ^ 2 + lgamma(a - 1) + log(pbdam(a, mult * m))
    vp <- exp(lcons) / vmax
    uvals <- c(-vp, vp)
  }
  # Acceptance-Rejection algorithm (ratio-of-uniforms)
  test <- FALSE
  counter <- 0
  max_iter <- 100
  while (!test) {
    u <- runif(1, uvals[1], uvals[2])
    v <- runif(1, 0, vmax)
    x <- (u / v) + mode
    test <- (2 * log(v) <= dgin1(x, a, m, TRUE, TRUE)) && (-mult * x > 0)
    counter <- counter + 1
    if (counter > max_iter) {
      x <- mode
      warning("No candidate draw found for these parameter values. Giving up an returning the mode of the distribution")
      break
    }
  }

  if (verbose) {
    return(list('value' = x, 'ARiters' = counter))
  } else {
    return(x)
  }
}

File: python_package/src/ginormal/utils/rtgin1.py
import numpy as np
from scipy import special, stats
from cardano_method import CubicEquation
import warnings

from ginormal.utils.mshift import mshift
from ginormal.utils.dgin1 import dgin1
from ginormal.utils.pbdam import pbdam

def rtgin1(a, m, sign, algo='hormann', verbose=False):
    # If sign = True, draw from the positive region Z > 0
    # If sign = False, draw from the negative region Z < 0
    if algo == 'hormann':
        algo_1 = True
    elif algo == 'leydold':
        algo_1 = False
    else:
        raise ValueError("algo_method must be either 'hormann' or 'leydold'")
    
    # Compute necessary values for ratio-of-uniforms method
    mult = 1 - 2*sign
    mode = (-m - mult*np.sqrt(m ** 2 + 4 * a)) / (2 * a)
    
    # Draw from minimal bounding rectangle and accept draw
    vmax = mshift(mode, mode, a, m, False)
    if algo_1:
        # Hörmann and Leydold (2014) using Cardano method
        roots = np.real(CubicEquation([2 - a, -m + a * mode, 1 + m * mode, -mode]).answers)
        roots = roots[-mult * roots > 0]
        uvals = np.sort([mshift(roots[j], mode, a, m) for j in range(2)])
    else:
        # Leydold (2001) using the proportionality constant
        lcons = -0.25 * m ** 2 + special.loggamma(a - 1) + np.log(pbdam(a, mult * m))
        vp = np.exp(lcons) / vmax
        uvals = np.array([-vp, vp])

    # Acceptance-Rejection algorithm (ratio-of-uniforms)
    test = False
    counter = 0
    max_iter = 100
    while (not test):
        u = stats.uniform.rvs(uvals[0], uvals[1] - uvals[0], size=1)[0]
        v = stats.uniform.rvs(0, vmax, size=1)[0]
        x = (u / v) + mode
        test = (2 * np.log(v) <= dgin1(x, a, m, True, True)) & (-mult * x > 0)
        counter += 1
        if counter > max_iter:
            x = mode
            warnings.warn("No candidate draw found for these parameter values. Giving up an returning the mode of the distribution")
            break
    if verbose:
        return {'value': x, 'ARiters': counter}
    else:
        return x
